File: typescript/matching-brackets/matching-brackets.ts
const Pairs: {[index: string]: string} = {
  '}': '{', ']': '[', ')': '('
}

class Stack<T> {
  private values: T[] = []
  push(value: T): void { this.values.unshift(value) }
  pop(): T | undefined { return this.values.shift() }
  isEmpty(): boolean   { return this.values.length === 0 }
}


export function isPaired(input: string): boolean {
  const stack = new Stack<string>()

  for (const ch of input) {
    if (Object.values(Pairs).includes(ch))   // open bracket
      stack.push(ch)

    else if (ch in Pairs)                    // close bracket
      if (stack.pop() !== Pairs[ch])
        return false
  }

  return stack.isEmpty()
}

File: dart/matching-brackets/lib/matching_brackets.dart
// map the close bracket to it's opening companion
const BRACKETS = <String,String>{
  ')': '(',
  ']': '[',
  '}': '{'
};

extension on String {
  bool get isOpenBracket => BRACKETS.values.contains(this);
  bool get isCloseBracket => BRACKETS.keys.contains(this);
  String openBracket() => BRACKETS[this] ?? '';
}

extension on List {
  void push<E>(E elem) => this.add(elem);
  E pop<E>() => this.removeLast();
}


class MatchingBrackets {
  bool isPaired(String input) {
    var stack = <String>[];

    for (var char in input.split('')) {
      if (char.isOpenBracket) {
        stack.push(char);
      }
      else if (char.isCloseBracket) {
        if (stack.isEmpty || char.openBracket() != stack.pop()) {
          return false;
        }
      }
    }
    return stack.isEmpty;
  }
}
